### PR TITLE
adding Static Constraint field in NumberProxy

### PR DESCRIPTION
**Pull request by jjsjann123**
## What does this PR do?

This PR fixes issue in: https://github.com/Lightning-AI/lightning-thunder/issues/463#issuecomment-2161242845

This PR does:
1. adding a `static_constraint` field on NumberProxy
2. marking `static_constraint` on NumberProxy args into `bool` method. The logic is behind that `bool` usually called on control flow args.
3. propagate `static_constraint` field in compute trace back to inputs, this should produce proper constraints in prologue.

TODO in future PRs:
1. We should be able to apply constant folding to simplify computation trace;
2. Finer grained constraints will encourage more cache re-use. But today we handle only static constraints where NumberProxy inputs needs to remain unchanged.

### CODE REVIEW

**Comment by jjsjann123** [comment]
Note for myself. @t-vi suggested in our offline discussion to have a forward propagation to specify which NumberProxy "can" be constrained. This would allow us to throw exception as we interpret python program, which would give better user signal.

I'll update this PR.

File: thunder/core/jit_ext.py [hunk 1]
```diff
@@ -1124,6 +1132,7 @@ def _general_jit_wrap_callback(value):
             pass
         elif should_register_for_prologue(value.provenance):
             value.provenance.ext_flag |= EXT_FLAG_IS_PROXY_DERIVED
+            value.provenance.ext_flag |= EXT_FLAG_IS_CONSTRAINABLE_INPUT
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply thunder/core/jit_ext.py:1135]: @t-vi wondering if this sounds about right to you?

[the remark this answers is not in the document]
Comment (jjsjann123) [reply thunder/core/jit_ext.py:1135]: Note that this is still a draft... It's functional but will need tests and comments before review.

**Comment by mruberry** [comment]
Not directly related to this PR, but a design review follow-up could be useful here. Doesn't have to be too crazy to prep, just an opportunity to outline and review the plan

**Comment by jjsjann123** [comment]
> Not directly related to this PR, but a design review follow-up could be useful here. Doesn't have to be too crazy to prep, just an opportunity to outline and review the plan

I'll follow up with the design review.
Sounds like @mruberry isn't blocking reviewing this PR until we clear the design review. In which case, this PR is ready for review.

File: thunder/core/jit_ext.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/core/jit_ext.py:1145]: Eh, sorry for the delay, yes. So there is a propagation rule for the ext flag in interpreter.py. If you don't need that, you don't need the flag, just as you describe.

**Review by t-vi — APPROVED** [review]
Looks good to me. Thank you @jjsjann123 . Let's see if the EXT_FLAG is useful for you and if not, remove it in a followup.

**Comment by csarofeen** [comment]
@jjsjann123 I’m a bit confused about what this PR does and where we’re going with this support. Whether or not we imminently do a design review, it would be valuable to write up the plan of attack for dynamic shapes, and better document what this PR does and how it fits into the overall plan. That could then be the basis for review.

**Comment by jjsjann123** [comment]
Yeah a design doc is over-due for dynamic shapes. I'll get some bandwidth on that once I wrap up the nvfuser work at hand.

Meanwhile, for this PR, its immediate usage is in supporting control flow. i.e. any control flow depending on a dynamic scalar means we should try to bake the scalar in as constant. I have some offline conversation with @t-vi on from this [thread](https://github.com/Lightning-AI/lightning-thunder/issues/463#issuecomment-2162643859).

Pinky promise that I'll get to a follow up write up about it in the dynamic shape design doc.

**Pull request merged**